File: motor_metrics/hold.py
"""Static-posture holds: sitting arms-free (GMFM dim B) and supported standing (dim D).

One function serves both. A supported stand and an unsupported sit are the same
measurement problem — how long, and how steadily — and the label carries the difference.

**What replaces the ordinal score.** The GMFM sitting items form a ladder of duration
thresholds (3 s, 5 s, 20 s, 60 s). That ladder is a continuous variable someone binned,
and re-reading the bin number back out would throw away the very resolution this package
exists to recover. So the headline metric is simply the **uncensored duration**, and
underneath it the **sway** of the trunk over the pelvis — the sub-clinical signal that
should move while the bin number sits still.

**Who decides what a hold is.** The annotator does. ``duration_s`` is the marked trial
length, because the in- and out-points *are* a human's judgment of when sitting began and
ended. Nothing here tries to infer loss-of-posture from a trunk-angle threshold: picking
that threshold would be inventing a clinical criterion and burying it in a constant. The
code's job starts inside the marked trial. ``tracked_s`` is a **data-quality** figure —
the longest stretch of continuously trusted tracking — and is not a claim about sitting.

**Two traps this closes.**

- ``path_length_m`` grows with trial length, so a *worse* 20-second hold beats a *better*
  8-second one on it. ``mean_velocity_mps`` is the duration-free form; ``window_s``
  truncates every trial to a common prefix; ``duration_s`` sits next to both so a
  length-confounded comparison is visible rather than inferred.
- ``hands_low_frac`` is **not** an arms-free detector. Whether a hand bears weight is a
  force question and there is no force sensor here. It is a QC flag for "your
  ``arms=free`` label may be wrong", and arms-free remains the annotator's assertion.
"""

from dataclasses import dataclass

import numpy as np

from .derive import resample_uniform, smooth
from .quality import TORSO, WRISTS, Gate, coverage, landmarks_ok, longest_run
from .signals import WORLD_UP, project_horizontal, trunk_from_vertical, trunk_vector

# Chi-square with 2 dof at p=0.95 — the standard posturography 95% sway ellipse.
_CHI2_95_2DOF = 5.991


@dataclass(frozen=True)
class HoldMetrics:
    """One static-hold trial. Distances in meters, angles in degrees, times in seconds."""

    duration_s: float  # the annotated trial: a human's call on when the hold began/ended
    tracked_s: float  # longest continuously-trusted run; a data-quality figure, not a hold
    coverage: float  # fraction of the trial with trusted torso landmarks
    n_frames: int

    # Sway of the trunk over the pelvis, in the plane perpendicular to `up`.
    path_length_m: float  # confounded with duration -- compare only at equal window_s
    mean_velocity_mps: float  # path_length / tracked_s; the duration-free form
    # 95% confidence ellipse of the 2D sway cloud. Read it next to the ML/AP split
    # below, never alone: a trunk rocking on a single axis traces a line, which
    # encloses no area, so this reads ~0 however hard it is rocking.
    ellipse_area_m2: float
    rms_m: float  # radial RMS excursion from the mean posture
    sway_ml_rms_m: float  # medio-lateral: in the image plane, measured well
    sway_ap_rms_m: float  # antero-posterior: inferred depth, markedly noisier

    trunk_angle_mean_deg: float  # unsigned lean from `up`; 0 = upright
    trunk_angle_sd_deg: float
    trunk_angle_range_deg: float

    hands_low_frac: float  # QC DIAGNOSTIC ONLY -- see the module docstring
    up_source: str  # which vertical was used; "world_y" assumes a level camera


def hold_metrics(
    rec,
    seg,
    *,
    up: np.ndarray = WORLD_UP,
    gate: Gate = Gate(),
    window_s: float | None = None,
) -> HoldMetrics:
    """Metrics for one ``sit_hold`` / ``stand_hold`` segment.

    Sway is computed over the longest continuously-tracked run inside the trial, never
    across a tracking gap — bridging one would invent the movement that happened inside
    it. Pass ``window_s`` to truncate that run to a common prefix so trials of unequal
    length can be compared on ``path_length_m``.

    Returns NaN sway fields (rather than raising) for trials that are empty, untracked,
    or shorter than the smoothing window. A mis-marked two-frame annotation is a normal
    thing to find in a session and must not take down a report with 40 good rows in it.
    """
    ts = np.asarray(rec.timestamps_ms)
    world = np.asarray(rec.landmarks_world)

    ok = landmarks_ok(rec, TORSO, gate)
    cov = coverage(ok, seg.start, seg.stop)
    duration_s = _span_seconds(ts, seg.start, seg.stop)

    r0, r1 = longest_run(ok, seg.start, seg.stop)
    if window_s is not None and r1 > r0:
        r1 = _clip_to_window(ts, r0, r1, window_s)
    tracked_s = _span_seconds(ts, r0, r1)

    sway = _sway(ts[r0:r1], world[r0:r1], up, tracked_s)
    angle = _trunk_angle_stats(world[r0:r1], up)

    return HoldMetrics(
        duration_s=duration_s,
        tracked_s=tracked_s,
        coverage=cov,
        n_frames=seg.stop - seg.start,
        **sway,
        **angle,
        hands_low_frac=_hands_low_frac(rec, r0, r1, up, gate),
        up_source="world_y" if np.allclose(up, WORLD_UP) else "custom",
    )


def sway_ellipse_area(points: np.ndarray) -> float:
    """95% confidence ellipse area of a 2D cloud; ``(N, 2)`` in, m^2 out.

    ``chi2(2, 0.95) * pi * sqrt(l1 * l2)`` over the covariance eigenvalues — the
    standard posturography figure. A perfectly collinear cloud has a zero eigenvalue and
    so zero area, which is correct: it encloses no region.
    """
    p = np.asarray(points, dtype=np.float64)
    if p.shape[0] < 3 or not np.isfinite(p).all():
        return float("nan")
    eigenvalues = np.linalg.eigvalsh(np.cov(p, rowvar=False))
    # Clamp: a degenerate cloud can produce a tiny negative eigenvalue from rounding.
    eigenvalues = np.clip(eigenvalues, 0.0, None)
    return float(_CHI2_95_2DOF * np.pi * np.sqrt(eigenvalues[0] * eigenvalues[1]))


def path_length(points: np.ndarray) -> float:
    """Total distance travelled along a ``(N, K)`` path."""
    p = np.asarray(points, dtype=np.float64)
    if p.shape[0] < 2 or not np.isfinite(p).all():
        return float("nan")
    return float(np.linalg.norm(np.diff(p, axis=0), axis=1).sum())


def _span_seconds(ts: np.ndarray, start: int, stop: int) -> float:
    if stop - start < 2:
        return 0.0
    return float(ts[stop - 1] - ts[start]) / 1000.0


def _clip_to_window(ts: np.ndarray, start: int, stop: int, window_s: float) -> int:
    """Shrink ``[start, stop)`` to the first ``window_s`` seconds of it."""
    cutoff = ts[start] + window_s * 1000.0
    return int(np.searchsorted(ts[start:stop], cutoff, side="right")) + start


def _sway(ts: np.ndarray, world: np.ndarray, up: np.ndarray, tracked_s: float) -> dict:
    """Sway of the trunk-over-pelvis path, on the pinned filter chain."""
    nan = dict(
        path_length_m=float("nan"),
        mean_velocity_mps=float("nan"),
        ellipse_area_m2=float("nan"),
        rms_m=float("nan"),
        sway_ml_rms_m=float("nan"),
        sway_ap_rms_m=float("nan"),
    )
    if world.shape[0] < 2:
        return nan

    horizontal = project_horizontal(trunk_vector(world), up=up)
    _, uniform = resample_uniform(ts, horizontal)
    if uniform.shape[0] == 0:
        return nan
    smoothed = smooth(uniform)  # NaN when the run is shorter than the window
    if not np.isfinite(smoothed).all():
        return nan

    centred = smoothed - smoothed.mean(axis=0)
    length = path_length(smoothed)
    return dict(
        path_length_m=length,
        mean_velocity_mps=length / tracked_s if tracked_s > 0 else float("nan"),
        ellipse_area_m2=sway_ellipse_area(smoothed),
        rms_m=float(np.sqrt((centred**2).sum(axis=1).mean())),
        sway_ml_rms_m=float(np.sqrt((centred[:, 0] ** 2).mean())),
        sway_ap_rms_m=float(np.sqrt((centred[:, 1] ** 2).mean())),
    )


def _trunk_angle_stats(world: np.ndarray, up: np.ndarray) -> dict:
    if world.shape[0] == 0:
        return dict(
            trunk_angle_mean_deg=float("nan"),
            trunk_angle_sd_deg=float("nan"),
            trunk_angle_range_deg=float("nan"),
        )
    angles = trunk_from_vertical(world, up=up)
    finite = angles[np.isfinite(angles)]
    if finite.size == 0:
        return dict(
            trunk_angle_mean_deg=float("nan"),
            trunk_angle_sd_deg=float("nan"),
            trunk_angle_range_deg=float("nan"),
        )
    return dict(
        trunk_angle_mean_deg=float(finite.mean()),
        trunk_angle_sd_deg=float(finite.std()),
        trunk_angle_range_deg=float(finite.max() - finite.min()),
    )


def _hands_low_frac(rec, start: int, stop: int, up: np.ndarray, gate: Gate) -> float:
    """Fraction of frames with either wrist below the pelvis, along ``up``.

    A hint that an ``arms=free`` label may be wrong, nothing more: a hand can rest low
    without bearing weight, and can bear weight without being low. Gated on the wrists
    separately from the torso, so a trial stays measurable when the hands are out of
    frame — this returns NaN there instead of poisoning the whole row.
    """
    if stop <= start:
        return float("nan")
    wrists_ok = landmarks_ok(rec, WRISTS, gate)[start:stop]
    if not wrists_ok.any():
        return float("nan")

    world = np.asarray(rec.landmarks_world)[start:stop]
    up_unit = np.asarray(up, dtype=np.float64)
    up_unit = up_unit / np.linalg.norm(up_unit)
    # Height above the pelvis (the world frame's origin) along `up`.
    heights = np.stack([world[:, i, :] @ up_unit for i in WRISTS], axis=1)
    return float(np.mean((heights[wrists_ok] < 0).any(axis=1)))

File: android/metrics/src/main/kotlin/dev/remapy/metrics/Hold.kt
package dev.remapy.metrics

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Static-posture holds: sitting arms-free (GMFM dim B) and supported standing (dim D).
 *
 * Port of `motor_metrics/hold.py`. One function serves both — a supported stand and an
 * unsupported sit are the same measurement problem, how long and how steadily, and the label
 * carries the difference.
 *
 * **What replaces the ordinal score.** The GMFM sitting items are a ladder of duration
 * thresholds (3 s, 5 s, 20 s, 60 s) — a continuous variable someone binned. So the headline
 * is the uncensored duration, and underneath it the **sway** of the trunk over the pelvis:
 * the sub-clinical signal that should move while the bin number sits still.
 *
 * **Two traps this closes**, both restated here because they are easy to lose in a port:
 * [HoldMetrics.pathLengthM] grows with trial length, so a *worse* 20-second hold beats a
 * *better* 8-second one on it — [HoldMetrics.meanVelocityMps] is the duration-free form. And
 * [HoldMetrics.ellipseAreaM2] reads ~0 for one-axis rocking, because a line encloses no area,
 * so it must be read next to the ML/AP split and never alone.
 *
 * Nothing here infers loss-of-posture from a trunk-angle threshold: picking that threshold
 * would be inventing a clinical criterion and burying it in a constant. Offline that job
 * belongs to the annotator; live there is no annotator, which is why the live path windows
 * on a clock instead and reports no `duration_s` at all.
 */
object Hold {

    /** Chi-square with 2 dof at p=0.95 — the standard posturography 95 % sway ellipse. */
    private const val CHI2_95_2DOF = 5.991

    /**
     * Metrics for one hold segment. Distances in meters, angles in degrees, times in seconds.
     *
     * Every measurement field is NaN rather than an exception for segments that are empty,
     * untracked, or shorter than the smoothing window. That is not defensive coding — a
     * mis-marked two-frame annotation is a normal thing to find in a session, and inside a
     * capture loop an unexpected throw costs the whole session rather than one table row.
     */
    data class HoldMetrics(
        /** The annotated trial: a human's call on when the hold began and ended. */
        val durationS: Double,
        /** Longest continuously-trusted run; a **data-quality** figure, not a hold. */
        val trackedS: Double,
        /** Fraction of the trial with trusted torso landmarks. */
        val coverage: Double,
        val nFrames: Int,
        /** Confounded with duration — compare only at equal `windowS`. */
        val pathLengthM: Double,
        /** `pathLength / trackedS`; the duration-free form. */
        val meanVelocityMps: Double,
        /** ~0 for one-axis rocking. Read beside the ML/AP split, never alone. */
        val ellipseAreaM2: Double,
        /** Radial RMS excursion from the mean posture. */
        val rmsM: Double,
        /** Medio-lateral: in the image plane, measured well. */
        val swayMlRmsM: Double,
        /** Antero-posterior: inferred depth, markedly noisier. */
        val swayApRmsM: Double,
        /** Unsigned lean from `up`; 0 = upright. */
        val trunkAngleMeanDeg: Double,
        val trunkAngleSdDeg: Double,
        val trunkAngleRangeDeg: Double,
        /** QC DIAGNOSTIC ONLY — see [handsLowFrac]. */
        val handsLowFrac: Double,
        /** Which vertical was used; `world_y` assumes a level camera. */
        val upSource: String,
    )

    /**
     * Metrics for one `sit_hold` / `stand_hold` segment.
     *
     * Sway is computed over the longest continuously-tracked run inside the trial, **never
     * across a tracking gap** — bridging one would invent the movement that happened inside
     * it. Pass [windowS] to truncate that run to a common prefix so trials of unequal length
     * can be compared on [HoldMetrics.pathLengthM].
     */
    fun holdMetrics(
        frames: PoseFrames,
        seg: Span,
        up: DoubleArray = Signals.WORLD_UP,
        gate: Quality.Gate = Quality.Gate(),
        windowS: Double? = null,
    ): HoldMetrics {
        val ok = Quality.landmarksOk(frames, Landmarks.TORSO, gate)
        val cov = Quality.coverage(ok, seg.start, seg.stop)
        val durationS = Quality.spanSeconds(frames, seg.start, seg.stop)

        var run = Quality.longestRun(ok, seg.start, seg.stop)
        if (windowS != null && run.stop > run.start) {
            run = Span(run.start, clipToWindow(frames, run, windowS))
        }
        val trackedS = Quality.spanSeconds(frames, run.start, run.stop)

        val sway = sway(frames, run, up, trackedS)
        val angle = trunkAngleStats(frames, run, up)

        return HoldMetrics(
            durationS = durationS,
            trackedS = trackedS,
            coverage = cov,
            nFrames = seg.nFrames,
            pathLengthM = sway.pathLengthM,
            meanVelocityMps = sway.meanVelocityMps,
            ellipseAreaM2 = sway.ellipseAreaM2,
            rmsM = sway.rmsM,
            swayMlRmsM = sway.swayMlRmsM,
            swayApRmsM = sway.swayApRmsM,
            trunkAngleMeanDeg = angle[0],
            trunkAngleSdDeg = angle[1],
            trunkAngleRangeDeg = angle[2],
            handsLowFrac = handsLowFrac(frames, run, up, gate),
            upSource = upSource(up),
        )
    }

    /** `world_y` when [up] is the default vertical, `custom` otherwise. */
    fun upSource(up: DoubleArray): String {
        // np.allclose's defaults: rtol 1e-5, atol 1e-8.
        val same = up.indices.all {
            abs(up[it] - Signals.WORLD_UP[it]) <= 1e-8 + 1e-5 * abs(Signals.WORLD_UP[it])
        }
        return if (same) "world_y" else "custom"
    }

    /**
     * 95 % confidence ellipse area of a 2D cloud, in m^2.
     *
     * `chi2(2, 0.95) * pi * sqrt(l1 * l2)` over the covariance eigenvalues — the standard
     * posturography figure. A perfectly collinear cloud has a zero eigenvalue and so zero
     * area, which is correct: it encloses no region.
     */
    fun swayEllipseArea(points: Matrix): Double {
        if (points.rows < 3 || !points.allFinite()) return Double.NaN
        val c = points.centred()
        val n = points.rows
        // np.cov's default normalisation is N - 1, not N.
        var sxx = 0.0
        var sxy = 0.0
        var syy = 0.0
        for (i in 0 until n) {
            sxx += c[i, 0] * c[i, 0]
            sxy += c[i, 0] * c[i, 1]
            syy += c[i, 1] * c[i, 1]
        }
        val denom = (n - 1).toDouble()
        val a = sxx / denom
        val b = sxy / denom
        val d = syy / denom
        // Symmetric 2x2 eigenvalues, ascending, as np.linalg.eigvalsh returns them.
        val mean = (a + d) / 2.0
        val radius = sqrt(((a - d) / 2.0) * ((a - d) / 2.0) + b * b)
        // Clamp: a degenerate cloud can produce a tiny negative eigenvalue from rounding.
        val low = (mean - radius).coerceAtLeast(0.0)
        val high = (mean + radius).coerceAtLeast(0.0)
        return CHI2_95_2DOF * PI * sqrt(low * high)
    }

    private class Sway(
        val pathLengthM: Double = Double.NaN,
        val meanVelocityMps: Double = Double.NaN,
        val ellipseAreaM2: Double = Double.NaN,
        val rmsM: Double = Double.NaN,
        val swayMlRmsM: Double = Double.NaN,
        val swayApRmsM: Double = Double.NaN,
    )

    /** Sway of the trunk-over-pelvis path, on the pinned filter chain. */
    private fun sway(frames: PoseFrames, run: Span, up: DoubleArray, trackedS: Double): Sway {
        if (run.nFrames < 2) return Sway()

        val horizontal = Signals.projectHorizontal(Signals.trunkVectors(frames, run), up)
        val (_, uniform) = Derive.resampleUniform(frames.timestampsMs(run.start, run.stop), horizontal)
        if (uniform.rows == 0) return Sway()
        val smoothed = Derive.smooth(uniform) // NaN when the run is shorter than the window
        if (!smoothed.allFinite()) return Sway()

        val centred = smoothed.centred()
        val length = pathLength(smoothed)
        var radial = 0.0
        var ml = 0.0
        var ap = 0.0
        for (i in 0 until centred.rows) {
            radial += centred[i, 0] * centred[i, 0] + centred[i, 1] * centred[i, 1]
            ml += centred[i, 0] * centred[i, 0]
            ap += centred[i, 1] * centred[i, 1]
        }
        val n = centred.rows
        return Sway(
            pathLengthM = length,
            meanVelocityMps = if (trackedS > 0) length / trackedS else Double.NaN,
            ellipseAreaM2 = swayEllipseArea(smoothed),
            rmsM = sqrt(radial / n),
            swayMlRmsM = sqrt(ml / n),
            swayApRmsM = sqrt(ap / n),
        )
    }

    /** Returns `[mean, sd, range]` of the trunk lean over the run, NaN-filtered. */
    private fun trunkAngleStats(frames: PoseFrames, run: Span, up: DoubleArray): DoubleArray {
        val nan = doubleArrayOf(Double.NaN, Double.NaN, Double.NaN)
        if (run.nFrames == 0) return nan
        val finite = Signals.trunkFromVertical(frames, run, up).filter { it.isFinite() }.toDoubleArray()
        if (finite.isEmpty()) return nan
        return doubleArrayOf(finite.average(), std(finite), finite.max() - finite.min())
    }

    /**
     * Fraction of frames with either wrist below the pelvis, along `up`.
     *
     * **Not an arms-free detector.** Whether a hand bears weight is a force question and there
     * is no force sensor here: a hand can rest low without bearing weight, and can bear weight
     * without being low. It is a hint that an `arms=free` label may be wrong, nothing more.
     *
     * Gated on the wrists separately from the torso, so a trial stays measurable when the
     * hands are out of frame — this returns NaN there instead of poisoning the whole row.
     */
    private fun handsLowFrac(
        frames: PoseFrames,
        run: Span,
        up: DoubleArray,
        gate: Quality.Gate,
    ): Double {
        if (run.stop <= run.start) return Double.NaN
        val wristsOk = Quality.landmarksOk(frames, Landmarks.WRISTS, gate)
        var considered = 0
        var low = 0
        val upUnit = Signals.unit(up)
        for (i in run.start until run.stop) {
            if (!wristsOk[i]) continue
            considered++
            val row = frames.worldRow(i)
            val anyLow = Landmarks.WRISTS.any { wrist ->
                val p = row.point(wrist)
                // Height above the pelvis (the world frame's origin) along `up`.
                p[0] * upUnit[0] + p[1] * upUnit[1] + p[2] * upUnit[2] < 0.0
            }
            if (anyLow) low++
        }
        if (considered == 0) return Double.NaN
        return low.toDouble() / considered
    }

    /** Shrink a run to the first [windowS] seconds of it. */
    private fun clipToWindow(frames: PoseFrames, run: Span, windowS: Double): Int {
        val cutoff = frames.timestampMs(run.start) + windowS * 1000.0
        // searchsorted(..., side="right"): first index whose timestamp exceeds the cutoff.
        var i = run.start
        while (i < run.stop && frames.timestampMs(i) <= cutoff) i++
        return i
    }
}
